Add "Configuration" support in UI for configuring the request timeout (and more things in the future)

pulkitsharma07 opened the pull request:
https://github.com/user-attachments/assets/ae6cc719-c31c-4fc4-9df3-ef9c3e7d539f

Fixes https://github.com/modelcontextprotocol/inspector/issues/142, removes support for setting timeout via search params (added as part of https://github.com/modelcontextprotocol/inspector/pull/164) , as that may not be intuitive to end users.

## Motivation and Context
Same as https://github.com/modelcontextprotocol/inspector/issues/142,

## How Has This Been Tested?
Tested locally by adding `console.log` statements when making a request in `useConnection` to a custom MCP server, was able to see the custom timeout value which was set manually in the UI, otherwise the default timeout of 10 seconds is used.

## Breaking Changes
Removes support for setting the request timeout value via search param: `timeout`

## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update

## Checklist
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [x] I have read the [MCP Documentation](https://modelcontextprotocol.io)
- [x] My code follows the repository's style guidelines
- [x] New and existing tests pass locally
- [x] I have added appropriate error handling
- [ ] I have added or updated documentation as needed

## Additional context
<!-- Add any other context, implementation notes, or design decisions -->

Cloud-Iris commented:
Yeah it is exactly what we need. Thanks for your work.🌹

olaservo commented on client/src/lib/constants.ts:
```diff
@@ -1,7 +1,17 @@
   CLIENT_INFORMATION: "mcp_client_information",
 } as const;
+
+export const DEFAULT_INSPECTOR_CONFIG: InspectorConfig = {
```
It could be helpful for future developers to add some clarification in the Readme (or elsewhere) to explain what types of configs belong in `InspectorConfig`.  Assuming my take on this matches your intentions, maybe something like:

```
# Configuration
The MCP Inspector supports two types of configurations:
- **Client configurations** apply universally to all MCP server connections and are managed through the Configuration UI.  Example: `MCP_SERVER_REQUEST_TIMEOUT` controls timeout duration for MCP server requests.
- **Server-dependent settings** are managed through other dedicated UI controls or through the CLI. Examples include `args` and environment variables.
```

Basically I am trying to set the stage for the incoming PR that adds full config file support for servers.

pulkitsharma07 replied:
I have gone through https://github.com/modelcontextprotocol/inspector/pull/177 and now I am thinking that we should have just a single config file (and remove the UI based configuration).
So the config file:
* configures which MCP server to use 
* and client specific configurations (like request timeout, authentication, proxy, error handling in the  future).

Wdyt ? I guess it depends on how people are using inspector, do they want to change config on the fly (via the UI) or they are okay changing the config file (and restarting inspector).

Personally, I am okay with a config file based approach, with nothing available in the UI. But not sure about other users, need your inputs here. Config file while need a restart and that can get rid of any "state" which MCP inspector had, so maybe we should go into UI direction ?

I think eventual solution would be to use config file as a base config, which is then overridden by the UI.

pulkitsharma07 replied:
Sure..

In terms of programmatic testing, I guess we will need a config file based approach for inspector config. I guess way forward is to get both PRs merged and then look into further improvements of combining them into a single config file.

olaservo commented on client/src/lib/hooks/useConnection.ts:
```diff
@@ -29,10 +29,6 @@ import { auth } from "@modelcontextprotocol/sdk/client/auth.js";
 import { authProvider } from "../auth";
 import packageJson from "../../../package.json";
 
-const params = new URLSearchParams(window.location.search);
```
Although the UI enhancement should eliminate the need for this, its technically a breaking change.  Should we leave in the option to set this through params?

pulkitsharma07 replied:
We can, but then I thought it should be really easy for someone to transition to the UI based config. 
Also based on discussions here I felt this may not be used that much, so just thought of removing it completely.

Let me know, if you still feel we should keep it, I can update the PR.

olaservo replied:
OK!  Its kind of an easter egg feature so I would be OK with removing this, and if there are any issues opened around it we can address those on a case by case basis.

olaservo reviewed (changes requested):
Thanks a lot for submitting this!  Had some minor feedback.

pulkitsharma07 commented on client/eslint.config.js:
```diff
@@ -19,6 +19,7 @@ export default tseslint.config(
     },
     rules: {
       ...reactHooks.configs.recommended.rules,
+      "react-hooks/rules-of-hooks": "off", // Disable hooks dependency order checking
```
I was getting a bunch of eslint warning in Cursor on opening the project for the first time. Particularly for [these](https://github.com/modelcontextprotocol/inspector/blob/4d4bb9110c5f4817e4c9dfb813b8fb217d2c3444/client/src/App.tsx#L65:L103) set of hooks, I believe that is happening because of [this](https://github.com/modelcontextprotocol/inspector/blob/4d4bb9110c5f4817e4c9dfb813b8fb217d2c3444/client/src/App.tsx#L55) condition.

Let me know if I need to configure something in Cursor to avoid the warning.

cliffhall replied:
The problem is that there is a conditional at the top of the component that short circuits the rest of the component by returning suspense if the path is oauth/callback. All the subsequent hooks complain because hooks are supposed to be run in the same order every time. Sometimes hooks would run, but sometimes we'd render the component as a suspense, bypassing them.

FYI: Fix is to move the conditional to the bottom where the render happens.
<img width="681" alt="Screenshot 2025-03-27 at 12 56 43 PM" src="https://github.com/user-attachments/assets/20ea0596-58bc-4dbc-a883-bff4372878d9" />

pulkitsharma07 replied:
ack, will fix it and turn it back on this PR.

cliffhall reviewed (changes requested):
Don't suppress rules of hooks. See comments. Will fully review when PR is ready to go again.

cliffhall reviewed (changes requested):
Looks good. Still agree with @olaservo that [we should have a note](https://github.com/modelcontextprotocol/inspector/pull/204#discussion_r2014521286) in the README about the configuration situation.

pulkitsharma07 commented:
I have added a [comment](https://github.com/pulkitsharma07/inspector/blob/9e1186f5ac8528c2947af2dd036c46cabfc458f8/client/src/lib/configurationTypes.ts#L6:L12) on `InspectorConfig` for developers  and added a [section](https://github.com/pulkitsharma07/inspector/tree/9e1186f5ac8528c2947af2dd036c46cabfc458f8?tab=readme-ov-file#configuration) in README for users.

Let me know how it looks @olaservo  @cliffhall ?

cliffhall reviewed (approved):
LGTM! 👍

cliffhall commented:
@pulkitsharma07  Tests for the Sidebar component are failing, breaking the build.
<img width="1185" alt="Screenshot 2025-03-29 at 12 24 54 PM" src="https://github.com/user-attachments/assets/5036e2b3-0b71-4ab7-848c-5ab84458a1be" />

cliffhall reviewed (changes requested):
Need to fix sidebar tests.

pulkitsharma07 commented:
@cliffhall fixed the tests, should I bump the version and update the screenshot as well ?

cliffhall reviewed (approved):
LGTM! 👍

cliffhall reviewed (changes requested):
Run prettier again, and it should be good build-wise.

A screenshot update seems like a good idea.

On the version number, I think we've tried to keep it in sync with the SDK mostly, but devoting [a specific PR](https://github.com/modelcontextprotocol/inspector/commit/16b38071e7554eb5556491f2f04339e58c5a073c) to bump.

pulkitsharma07 commented:
@cliffhall  ah, sorry should've run it locally first. I have updated the contributions guidelines for the same.

cliffhall commented:
@pulkitsharma07 Did you want to replace the screenshot?

pulkitsharma07 commented:
@cliffhall  updated https://github.com/pulkitsharma07/inspector/blob/main/mcp-inspector.png

cliffhall reviewed (approved):
LGTM! 👍

olaservo commented:
Hi @cyri113 still working on getting the latest changes published to npm. It should happen within the next couple days and I can follow up here once its live.
